### ISSUE TITLE
IList sort for Dart 2 strong mode issues

### ISSUE THREAD

**Issue by rich-j** [spebbe/dartz#5]
I'm encountering 2 issues when sorting an `IList` in Dart 2 running in strong mode on DDC.

The first issue is a runtime `CastError` on the `IList`  `partition` function which is called by the `sort` function:

    CastError: Casting value of type 'Tuple2<IList, IList>' to type 'Tuple2<IList<Person>, IList<Person>>' which is incompatible in strong mode

I was able to get it to work by overriding your `IList` `partition` implementation with:

    Tuple2<IList<A>, IList<A>> partition(bool f(A a)) =>
      foldRight(tuple2(empty(), empty()), (A a, acc) => f(a)
        ? acc.map1((xs) => xs.prependElement(a))
        : acc.map2((xs) => xs.prependElement(a)));

I looked into changing the type parameters for `TraversableMonadPlusOps<F, A>` from `<IList, A>` to `<IList<A>, A>` but was quickly over my head in dependencies. This is probably covered by your work for https://github.com/spebbe/dartz/issues/2. Any timeframe for supporting sort?

The second issue is probably not directly a dartz issue. The `IList` `sort` requires an `Order` object to provide the object comparison functions. We receive this exception at runtime:

    EXCEPTION: Type 'PersonNameOrder' is not a subtype of type 'Order<PersonObj>'

Our app has a common library used by various U/I layers. The common API provides abstract classes that are implemented in private common code. The U/I doesn't have access to the implementation classes so can only specify the sort order using the API classes.

    class Person { String get name;  }    // API
    class PersonObj implements Person { ... }    // private to U/I

    class PersonNameOrder extends Order<Person> { ... }

    IList<Person> persons = ...   // Instantiated in the private layer as an IList<PersonObj>
    persons.sort( new PersonNameOrder() );

It's more complex that this but there seems to be an issue with variance. PersonObj is a subtype of Person so is covariant. Providing the ordering function should then be contravariant. How is this specified in Dart?

I have been able to get this to work:

    persons.map((p)=>p).sort( new PersonNameOrder() );
  
Note the identity mapping function that casts each `PersonObj` to a `Person`. Any insights or other thoughts to correctly specify a sort?

**Comment by spebbe** [comment]
Sorry about the long wait -- I missed this one for some reason.

Thanks for reporting! For now, I've pushed a workaround for IList sort that is very similar to yours on the ddc_compat branch. I'd be happy to give you more explicit credit if you want. I'm still hoping to find some way of expressing this in a less repetitious way once Dart 2 has stabilized.

Regarding variance in Dart: Generics are always covariant in Dart whether you want it or not, even though this is unsound for mutable structures. There have been a couple of tweaks to how in-/co-/contravariance is specified and interpreted on the road to Dart 2, but the setup is still very limiting. I believe the recently introduced 'retype' methods on List, Stream, etc are the official workarounds for issues similar to what you are seeing and serve roughly the same purpose as your 'map((p) => p)'.

Here's an alternative workaround that is subjectively less gross:
```dart
class PersonNameOrder<Impl extends Person> extends Order<Impl> {
  @override Ordering order(Impl a1, Impl a2) => StringOrder.order(a1.name, a2.name);
}

...

persons.sort(new PersonNameOrder<PersonObj>());
```

Hope that helps!

**Comment by rich-j** [comment]
Thanks for publishing a fix. The credit is all yours. I included my fix mostly for completeness of describing the problem. I know that you are in a tough position waiting for Dart 2 to stabilize.

The secondary issues of specifying sort order I handled by creating a set of singleton ordering objects. The information you provided in your answer to issue #6 suggests that ordering singletons are good for equality checks and also solved my variance issue.
